Evaluation/evaluate_performance.py

```diff
@@ -0,0 +1,174 @@
+import os
+import math
+import json
+from collections import defaultdict
+from dataclasses import fields
+
+import mlflow
+import hydra
+from hydra.utils import to_absolute_path
+from omegaconf import OmegaConf, DictConfig, ListConfig
+
+import eval_tools
+
+@hydra.main(config_path='configs', config_name='run3')
+def main(cfg: DictConfig) -> None:
+    mlflow.set_tracking_uri(f"file://{to_absolute_path(cfg.path_to_mlflow)}")
+    
+    # setting paths
+    path_to_input_taus = to_absolute_path(cfg.path_to_input_taus) if cfg.path_to_input_taus is not None else None
+    path_to_input_vs_type = to_absolute_path(cfg.path_to_input_vs_type) if cfg.path_to_input_vs_type is not None else None
+    path_to_pred_taus = to_absolute_path(cfg.path_to_pred_taus) if cfg.path_to_pred_taus is not None else None
+    path_to_pred_vs_type = to_absolute_path(cfg.path_to_pred_vs_type) if cfg.path_to_pred_vs_type is not None else None
+    path_to_target_taus = to_absolute_path(cfg.path_to_target_taus) if cfg.path_to_target_taus is not None else None
+    path_to_target_vs_type = to_absolute_path(cfg.path_to_target_vs_type) if cfg.path_to_target_vs_type is not None else None
+    path_to_weights_taus = to_absolute_path(cfg.path_to_weights_taus) if cfg.path_to_weights_taus is not None else None
+    path_to_weights_vs_type = to_absolute_path(cfg.path_to_weights_vs_type) if cfg.path_to_weights_vs_type is not None else None
+    #
+    path_to_artifacts = to_absolute_path(f'{cfg.path_to_mlflow}/{cfg.experiment_id}/{cfg.run_id}/artifacts/')
+    output_json_path = f'{path_to_artifacts}/performance.json'
+
+    # init Discriminator() class from filtered input configuration
+    field_names = set(f_.name for f_ in fields(eval_tools.Discriminator))
+    init_params = {k:v for k,v in cfg.discriminator.items() if k in field_names}
+    discriminator = eval_tools.Discriminator(**init_params)
+    
+    # init PlotSetup() class from filtered input configuration
+    field_names = set(f_.name for f_ in fields(eval_tools.PlotSetup))
+    init_params = {k:v for k,v in cfg.plot_setup.items() if k in field_names}
+    plot_setup = eval_tools.PlotSetup(**init_params)
+
+    # construct branches to be read from input files
+    read_branches = OmegaConf.to_object(cfg.read_branches)
+    if discriminator.from_tuple:
+        read_branches.append(discriminator.pred_column)
+        if discriminator.wp_column != discriminator.pred_column:
+            read_branches.append(discriminator.wp_column)
+
+    # read original data with corresponging predictions into DataFrame
+    if path_to_input_vs_type is None:
+        df_all = eval_tools.create_df(path_to_input_taus, read_branches, path_to_pred_taus, path_to_target_taus, cfg.discriminator.pred_column_prefix, path_to_weights_taus)
+        # apply gen. selection
+        inclusive_gen_selection = ' or '.join([f'(gen_{tau_type}==1)' for tau_type in ['tau', cfg.vs_type]]) # gen_* are constructed in `add_targets()`
```

Comment:
For example, if I want to evaluate roc for "taus from H->tautau" vs "jets from DY", at which point it is selected that taus should be taken from "H->tautau" and jets from "DY"? At this point, two datasets are already merged, so such selection should happen before, but then why it is also needed here?

Reply:
Actually, the behaviour you described is the one assumed by default, but maybe it is not articulated clearly in the code. The "if" case which you referenced above I inherited from the previous code, and I think the logic behind it was to cover the case when both taus and jets are taken from the same single source. 

So setting `path_to_input_vs_type=None` triggers this case and both taus and jets are taken from `path_to_input_taus`, with predictions taken from `path_to_pred_taus ` and targets from `path_to_target_taus `. Then, one can apply this inclusive OR selection in order to select both types.

On the contrary, if one wants to have taus from HTT and jets from DY, `path_to_input_vs_type` should be specified to DY file and then it triggers the "else" case, which takes taus from `path_to_input_taus`, other type from `path_to_input_vs_type` and applies gen matching separately:
```python
    else:
        df_taus = eval_tools.create_df(path_to_input_taus, read_branches, path_to_pred_taus, path_to_target_taus, cfg.discriminator.pred_column_prefix, path_to_weights_taus)
        df_vs_type = eval_tools.create_df(path_to_input_vs_type, read_branches, path_to_pred_vs_type, path_to_target_vs_type, cfg.discriminator.pred_column_prefix, path_to_weights_vs_type)
        df_taus = df_taus.query('gen_tau==1')
        df_vs_type = df_vs_type.query(f'gen_{cfg.vs_type}==1')
        df_all = df_taus.append(df_vs_type)
```

I was working on it for quite a long time so I got used to this logic, but not sure now that it is the most intuitive one:)

Reply:
I see. So it is some merge of new and old approaches... I think it would be preferable that we have two yaml configs one to run prediction other to run plots and then have tau type filters specified there. In this way, in the code, if/else wouldn't be necessary.
E.g. in the prediction yaml for DY we want to run predictions for all types (e,mu,tau,jet). While for rocs one can want to select (tau from HTT vs jet from DY) or (tau from DY vs jet from DY), or some crazy scenarios like (a mix of taus from HTT and DY vs a mix of jets from W and TT) etc. I.e. in the roc config, we don't need to define separate inputs "taus" and "other", but rather have a set of entries (file_name_pattern, tau_selection to apply to it).

Reply:
> I.e. in the roc config, we don't need to define separate inputs "taus" and "other", but rather have a set of entries (file_name_pattern, tau_selection to apply to it).

@kandrosov Well, I think this logic can't be implemented at the `plot_roc.py` step, but can be (and in fact is already) included at `evaluate_performance.py` step.  In general, the idea of `plot_roc.py` is that it just extracts an entry from `performance.json` based on `sample` labels (where `tau` and `vs_type` come from) and plots them, so all the selections/roc computation need to be performed at the earlier stage.

But as I mentioned various scenarios can be realised and dumped into `performance.json` already at `evaluate_performance.py` step. `apply_training.py` requires just _some_ input ROOT file to produce predictions, so it can be e.g. `DY=jets+mu+ele` mixture, or already some preselected/mixture-of-samples file. 

Since it is `evaluate_performance.py` where ROC is computed, it will _always_ require two inputs (labelled as `*_taus` and `*_vs_type`) - but there is a freedom here to pick those inputs. The only assumption which is made is that on one will be applied the selection `gen_tau==1`, on the other `gen_{vs_type}==1` and a common cut on their combination. Then the ROC is computed and is stored in `performance.json` with identifiers `sample_tau/sample_vs_type/vs_type/pt_bin`. Based on this identifiers, it is then retrieved and plotted by `plot_roc.py`.

That means, that one can pass the same DY input file to `*_taus` and `*_vs_type` arguments (input/predictions/targets) to take both taus and jets from it. Or, inputs can be DY for taus and HTT for jets. As for the mixture case (e.g. W+TT for jets, HTT+DY for taus), this input needs to be constructed manually beforehand (e.g. `hadd`-ed) and be presented as two tuples of _single_ files with (h5 with predictions/ h5 with targets/ h5 with weights/ ROOT with inputs) one per tau/vs_type to `evaluate_performance.py` with a corresponding label `sample_vs_type=W_TT`, `sample_tau=HTT_DY`. Then in `plot_roc` one can plot this ROC curve by providing those labels.

Hopefully, I managed to explain the idea more or less clearly. So finally, coming back to the original "if/else" discussion, I think "if" case can be simply removed, leaving the "else" (two single inputs) case left. The variability of scenarios can then be included as I described above. Would this logic make sense?

Reply:
thanks for the detailed reply @yaourtpourtoi ! I probably still didn't understand the treatment of `gen_tau` and `vs_type` datasets should be done differently. I don't particularly like the idea of using `hadd`. Wouldn't it be more ergonomic to define samples in `run3.yaml` in this way: 
```yaml
samples:
  DY: [ 'tau' ]
  TT: [ 'jet' ]
  W: [ 'tau', 'jet' ]
  ...
path_to_input: 'eval_data/${sample}*.root'
path_to_pred: '${path_to_mlflow}/${experiment_id}/${run_id}/artifacts/predictions/${sample}*.h5'
...
```
and then in `evaluate_performance.py`: simply loop over all files an apply corresponding selection. In this way, also multiple files for the single sample are supported without requiring special treatement.

Reply:
Ah, I see, this is indeed an elegant way to treat multiple inputs - will add it to this PR